src/lib/salad/hll.c

```diff
@@ -0,0 +1,223 @@
+{
+	size_t first_byte_pos = idx * HLL_REGISTER_BITS / 8;
+	bucket->addr = (uint16_t *)(regs + first_byte_pos);
+	bucket->offset = idx * HLL_REGISTER_BITS - first_byte_pos * 8;
+}
+
+/* Get the value of the register with the idx index. */
+static inline uint8_t
+hll_register_rank(const struct hll *hll, size_t idx)
+{
+	struct reg_bucket bucket;
+	init_reg_bucket(&bucket, hll->registers, idx);
+	uint16_t reg_mask = hll_ones(HLL_REGISTER_BITS);
+	uint16_t bucket_val = *bucket.addr;
+	uint8_t rank = (bucket_val >> bucket.offset) & reg_mask;
+	return rank;
+}
+
+/* Set rank as the new value of the register with the idx index. */
+static inline void
+hll_set_register_rank(struct hll *hll, size_t idx, uint8_t rank)
+{
+	struct reg_bucket bucket;
+	init_reg_bucket(&bucket, hll->registers, idx);
+	uint16_t boundary_mask =
+		~(hll_ones(HLL_REGISTER_BITS) << bucket.offset);
+	uint16_t bucket_val = *bucket.addr;
+	*bucket.addr = (rank << bucket.offset) | (bucket_val & boundary_mask);
+}
+
+void
+hll_add(struct hll *hll, uint64_t hash)
+{
+	uint8_t precision = hll->precision;
+	size_t idx = hll_hash_register_idx(hash, precision);
+	assert(idx < hll_n_registers(precision));
+	uint8_t hash_rank = hll_hash_rank(hash, precision);
+	uint8_t reg_rank = hll_register_rank(hll, idx);
+	if (reg_rank < hash_rank) {
+		hll_set_register_rank(hll, idx, hash_rank);
+		hll_invalidate_cache(hll);
+	}
+}
+
+/* Alpha constant that HyperLogLog uses for estimating. */
+static inline double
+hll_alpha(uint8_t precision)
+{
+	return 0.7213 / (1.0 + 1.079 / hll_n_registers(precision));
+}
+
+/*
+ * Estimate the cardinality using the HyperLogLog estimation formula.
+ * Raw estimation can have larger relative error for small cardinalities.
+ */
+static double
+hll_raw_estimate(const struct hll *hll)
+{
+	double sum = 0;
+	const size_t n_registers = hll_n_registers(hll->precision);
+	for (size_t i = 0; i < n_registers; ++i) {
+		sum += pow(2, -hll_register_rank(hll, i));
+	}
+
+	const double alpha = hll_alpha(hll->precision);
+	return alpha * n_registers * n_registers / sum;
+}
+
+/* Count the number of registers that are zero. */
+static size_t
+hll_count_zero_registers(const struct hll *hll)
+{
+	size_t count = 0;
+	const size_t n_registers = hll_n_registers(hll->precision);
+	for (size_t i = 0; i < n_registers; ++i) {
+		if (hll_register_rank(hll, i) == 0)
+			++count;
+	}
+	return count;
+}
+
+/* Estimate the cardinality using the LinearCounting algorithm. */
+static inline double
+linear_counting(size_t regs, size_t zero_count)
+{
+	return regs * log((double)regs / zero_count);
+}
+
+uint64_t
+hll_estimate(const struct hll *hll)
+{
+	if (hll_is_valid_cache(hll->cached_estimation)) {
+		return hll->cached_estimation;
+	}
+	const uint8_t prec = hll->precision;
+	const size_t n_registers = hll_n_registers(prec);
+	double raw_estimation = hll_raw_estimate(hll);
+
+	double hll_estimation = raw_estimation;
+	if (raw_estimation < 5 * n_registers) {
```

Comment:
Please remove magic number `5`.  Isn't supposed to be `5 * n_registers / 2`?

Reply:
It's not completely clear which border we should use. Google developers state that after 5m there is no need for a correction so that is the boundary that used in the article. Redis developers achieved the best accuracy for the 3m boundary. I'm currently working on test for HyperLogLog and I got the best results for 4m.